pubsub/README.rst

```diff
@@ -139,6 +139,25 @@ any metrics agent that implements the same interface as ``MetricsAgent``
   in a single request.
 
 
+If you prefer pull-based metrics like Prometheus you will be pleased to know that
+the subscriber records Prometheus metrics of the form ``gcloud_aio_<metric>``,
+which will be ignored if you don't use Prometheus and will be automatically
+included if you do:
+
+- ``pubsub_consume_total`` (labels: ``outcome = {'succeeded', 'cancelled',
+  'failed', 'failfast'}``) - [counter] a consume operation has completed with a
+  given outcome
+- ``pubsub_consume_latency_seconds`` (labels: ``aspect = {'receive',
+  'runtime'}``) - [histogram] how many seconds taken to receive a message, or
+  callback runtime
+- ``pubsub_batch_total`` (labels: ``component = {'acker', 'nacker'}, outcome =
+  {'succeeded', 'failed'}``) - [counter] a batch has succeeded or failed to be
+  acked or nacked
+- ``pubsub_messages_processed_total`` (labels: ``component = {'acker',
+  'nacker', 'producer'}``) - [counter] the number of messages that were acked,
+  nacked, or retrieved from pubsub
+
+
```

Comment:
I think the above usages of labels vs metrics all make sense. Fundamentally, each metric has semantic meaning as far as I call tell: eg. `sum(pubsub_consume_total) by (outcome)` is something that would be absolutely useful to track on its own and to view the sratio of the different `outcome` values.

My only counter-example here would be `pubsub_consume_latency_seconds` -- as far as I can tell, there's some time unaccounted for there (between when we spawn the callback task and when it actually starts) -- we might need a `pubsub_consume_latency_seconds, aspect = "asyncio_waiting"` (naming tbd) for the fully aggregated metric to be entirely meaningful. Thoughts?

(Note: I actually _really like_ how you grouped these metrics for exactly the reason of making the above obvious. I hadn't even realized that there was time being unaccounted for in our metrics until I went to check on whether these metrics are fully meaningful as a sum across all labels!)

Reply:
I also really like the naming of your labels -- `aspect`, `component`, and `outcome` might be nice to standardize across our other PRs.

Reply:
curious why'd you pick `aspect`? I associate that with multiple perspectives on the same object. So like I'd assume it mean something like the 'reliability aspect' or 'latency aspect' whereas here its trying to label separate parts of a sequence... right?

how about `one_of([step, stage, phase])`

love `outcome` lets do that

Reply:
I do definitely like `step` for something that has an implied ordering (eg. perfect for etl, for example). Less certain about it for places that aren't ordered, though I don't believe we've hit that case yet. I'm ok with `phase` as well.

Veto for `stage`, since we already use that as the `[canary, release, pre-release, ...]` label.

Reply:
**re: Kevin**

Ah, good point, it would be good to measure the "queueing" time so to speak.

**re: Allan**

I picked `aspect` as it describes a different "angle" on the same metric and it provides a space to insert a descriptor for how the metric differs from others with the same name (ex. one aspect of latency is how long it took to process the thing, one aspect is how long it took for the thing to start being processed, one aspect could be the sum of the two, etc.).

I'm happy to choose any other name for `aspect` although like Kevin mentioned, `step` implies ordering which I don't aim to achieve with `aspect`. I would also argue that `stage` and `phase` also imply a kind of non-overlapping sequence of measurements, but that interpretation could just be my own bias. Although we could just force this to be a sequential, non-overlapping metric and use one of the terms you suggested instead. Thoughts?

Reply:
> what do you think about including a "total" time including receive, runtime, async queueing, and ack time which we could use to calculate out the non-obvious processing times?

We might have to play around with this a bit to see how well it works. I think for aggregated metrics, it's always better to "add up" to a total instead of subtracting from it, though that might be bad memories of Datadog bleeding through for me rather than something more globally accurate.

Reply:
I would submit to you there may now be merit in providing a timestamp to perform subtractive operations on opposed to additive ones (if only to record the perceived time based from a user-provided timestamp).

Alternatively, we could utilize Google Cloud Spanner for high fidelity measurements to more precisely gauge our system and its performance metrics: https://cloud.google.com/spanner/docs/true-time-external-consistency.

It's been a while, but I studied Spanner in "Designing Data Intensive Applications".

Reply:
Interesting find @caseydialpad, I wonder if the Cloud Spanner TrueTime is applicable to nodes outside of Cloud Spanner? It looks like it uses atomic clocks as well as only allowing transactions on intervals. I found this explainer blog and this is an interesting block:

> It’s important to keep in mind that TrueTime does not guarantee perfectly synchronized clocks. Rather, TrueTime gives an upper bound for clock offsets between nodes in a cluster. Synchronization hardware helps minimize the upper bound. In Spanner’s case, Google mentions an upper bound of 7ms. That’s pretty tight; by contrast, using [NTP](https://en.wikipedia.org/wiki/Network_Time_Protocol) for clock synchronization is likely to give somewhere between 100ms and 250ms.

Reply:
I see what Allan is saying about `step` versus `aspect`. Formerly there were `receiving` and `runtime` that could be interpreted as steps or aspects. They could be viewed as steps if they are intended to be separate stages, but they could also be aspects if we say `runtime` includes `receiving`. I think the latter muddies intuitive interpretation.

Perhaps I would ask what the DE definitions of `aspect` and `step` are so we can consistently use them.

To me the intuitive use of aspect would be to in terms of perception as in the case of art, and user experience. The user-aspect would be metrics that reflect how a user perceives something. Another aspect might be how a speaker with a non-optimal decoder perceives transcription. Another still might be transformations on the data to reflect some system aspect of the system (such as the memory requirements of the chunks of audio transcription)  to help us perceive what the decoder experiences as it has different experiences digesting information.

For steps, the term to me denotes something more discrete with beginning and ends as opposed to sequences. A sequence is definitely composed of steps but steps may be composed of substeps (so we'd need to tighten up the terminology). Like a CPU many things may happen in parallel, but each step is discrete and exclusive to itself. I'd also argue steps are less a matter of perception and more a matter of impersonal measurement.

Returning to `runtime`, `queue` and `receiving`, these feel strongly of `steps` given the discreteness of their nature. Each have well defined beginning and ends.

Reply:
@jonathan-johnston, you're right. If I recall correctly, the timestamp received is some interval with a degree of confidence attached to it (>95% I think). I've never used it, but where high quality timing is required and no internally consistent reference is available, the existence of remote atomic clocks are about as good as it gets.

Reply:
Thanks for your thoughtful post Casey! After reading your post I feel like:

* `step` should be some kind of verb, and describes something that you have control over (like ETL where extract, transform, and load are all discrete steps which the code in question _actively does_) and of course is a series of steps which achieves some goal
* `phase` can be something that is happening which generally progresses toward a final state but it may not be something that the code in question _does_ (ex. it does not _ actively do_ `runtime` but phrased another way it could _actively do_ `processing` or `process`)
* `aspect` is as discussed in that it describes a facet of something and doesn't have to be sequential in nature

Overall I am leaning toward `phase` or `aspect` for `receive`, `queuing`, and `runtime` of the `consume_latency` metric.

I hope I'm interpreting this correctly, but it looks to me like there are no strong opinions for which word we should use. I feel like `step` should definitely be used for measurement of actively executed steps (ex. ETL) and `aspect` should be used for non-actively executed things (ex. `receive`).

Reply:
Those definitions make sense to me, I'd be comfortable using the above as our guidelines. Unless there are any strong objections, let's go with how Jon laid out those descriptions (and @jonathan-johnston could you please reflect this in https://docs.talkiq-echelon.talkiq.com/static/docs/guides/observability/metrics.html#common-names ?)

Reply:
@TheKevJames Added the recommendations here: https://github.com/talkiq/docs/pull/239

Reply:
FYI I have settled on `phase` for this metric since the `consume_latency` metric is progressing toward and "end state" of the incoming message being completely consumed but the metric is making passive observations (therefore, `step` is out of the question).